src/rules_dir_watcher.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This module implements custom Juju event (alert_rules_changed) fired upon any change in a given
+directory mounted to the workload container. It is based on `watchdog`.
+In this particular case, it is used by the prometheus-configurer-k8s-operator charm to detect
+changes of the alerting rules. Thanks to this mechanism, Prometheus Configurer knows when to
+update the configuration of the Prometheus Server.
+"""
+
+import logging
+import os
+import subprocess
+import sys
+import time
+
+from ops.charm import CharmBase, CharmEvents
+from ops.framework import EventBase, EventSource, Object
+from watchdog.events import FileSystemEventHandler  # type: ignore[import]
+from watchdog.observers import Observer  # type: ignore[import]
+
+logger = logging.getLogger(__name__)
+
+
+class AlertRulesChangedEvent(EventBase):
+    pass
+
+
+class AlertRulesChangedCharmEvents(CharmEvents):
+    alert_rules_changed = EventSource(AlertRulesChangedEvent)
+
+
+LOG_FILE_PATH = "/var/log/prometheus-configurer-watchdog.log"
+
+
+class AlertRulesDirWatcher(Object):
+    def __init__(self, charm: CharmBase, rules_dir: str):
+        super().__init__(charm, None)
+        self._charm = charm
+        self._rules_dir = rules_dir
+
+    def start_watchdog(self):
+        """Starts watchdog."""
```

Comment:
There is this docstring plus another one that says "Starts the actual watchdog process.", it's a bit confusing

Reply:
True. Hope new docstrings are better.